Makefile

```diff
@@ -1,21 +1,91 @@
-RETOOL=$(CURDIR)/_tools/bin/retool
-PATH := ${PWD}/bin:${PWD}/ENV/bin:${PATH}
+SHELL := /bin/bash -o pipefail
+UNAME_OS := $(shell uname -s)
+UNAME_ARCH := $(shell uname -m)
+
+TMP_BASE := .tmp
+TMP := $(TMP_BASE)/$(UNAME_OS)/$(UNAME_ARCH)
+TMP_BIN = $(TMP)/bin
+
+DEP_VERSION := 0.5.0
+DEP := $(TMP_BIN)/dep
+ifeq ($(UNAME_OS),Darwin)
+DEP_OS := darwin
+endif
+ifeq ($(UNAME_OS),Linux)
+DEP_OS = linux
+endif
+ifeq ($(UNAME_ARCH),x86_64)
+DEP_ARCH := amd64
+endif
+$(DEP):
+	@mkdir -p "$(TMP_BIN)"
+	curl -sSL "https://github.com/golang/dep/releases/download/v$(DEP_VERSION)/dep-$(DEP_OS)-$(DEP_ARCH)" -o "$(DEP)"
+	@chmod +x "$(DEP)"
+
+PROTOC_VERSION := 3.7.0
+PROTOC := $(TMP_BIN)/protoc
+ifeq ($(UNAME_OS),Darwin)
+PROTOC_OS := osx
+endif
+ifeq ($(UNAME_OS),Linux)
+PROTOC_OS = linux
+endif
+ifeq ($(UNAME_ARCH),x86_64)
+PROTOC_ARCH := x86_64
+endif
+$(PROTOC):
+	@mkdir -p "$(TMP_BIN)"
+	curl -sSL "https://github.com/protocolbuffers/protobuf/releases/download/v$(PROTOC_VERSION)/protoc-$(PROTOC_VERSION)-$(PROTOC_OS)-$(PROTOC_ARCH).zip" -o "$(TMP)/protoc.zip"
+	cd "$(TMP)"; unzip protoc.zip
+	@rm -f "$(TMP)/readme.txt"
+
+ERRCHECK_VERSION := v1.2.0
+ERRCHECK := $(TMP_BIN)/errcheck
+$(ERRCHECK):
+	$(eval ERRCHECK_TMP := $(shell mktemp -d))
+	@cd $(ERRCHECK_TMP); echo module tmp > go.mod; GO111MODULE=on go get github.com/kisielk/errcheck@$(ERRCHECK_VERSION)
+	@rm -rf $(ERRCHECK_TMP)
+
+PROTOC_GEN_GO_VERSION := v1.3.0
+PROTOC_GEN_GO := $(TMP_BIN)/protoc-gen-go
+$(PROTOC_GEN_GO):
+	$(eval PROTOC_GEN_GO_TMP := $(shell mktemp -d))
+	@cd $(PROTOC_GEN_GO_TMP); echo module tmp > go.mod; GO111MODULE=on go get github.com/golang/protobuf/protoc-gen-go@$(PROTOC_GEN_GO_VERSION)
+	@rm -rf $(PROTOC_GEN_GO_TMP)
+
+PROTOC_GEN_GOFAST_VERSION := v1.2.1
+PROTOC_GEN_GOFAST := $(TMP_BIN)/protoc-gen-gofast
+$(PROTOC_GEN_GOFAST):
+	$(eval PROTOC_GEN_GOFAST_TMP := $(shell mktemp -d))
+	@cd $(PROTOC_GEN_GOFAST_TMP); echo module tmp > go.mod; GO111MODULE=on go get github.com/gogo/protobuf/protoc-gen-gofast@$(PROTOC_GEN_GOFAST_VERSION)
+	@rm -rf $(PROTOC_GEN_GOFAST_TMP)
+
+export GOBIN := $(abspath $(TMP_BIN))
```

Comment:
This aliased GOBIN seems unneeded, `export PATH := $(abspath $(TMP_BIN)):$(PATH)` seems just as clear. But again minor.

Reply:
It's needed so that the `go get` and `go install` calls install to the correct location, it makes it so that `errcheck, protoc-gen-go, protoc-gen-gofast, protoc-gen-twirp, protoc-gen-twirp_python` all install to this directory.